### PR TITLE
Support mxint4 flashinfer_trtllm moe gemm

### PR DESCRIPTION
**Pull request by HandH1998**
## Motivation

We introduced an MXINT4 flashinfer_trtllm MoE GEMM kernel on SM100. We report both accuracy and performance results of the Kimi-K2-Thinking model on B200.  Compared with Marlin MoE, the new kernel delivers higher prefill throughput, and improves decoding throughput for most batch sizes. This kernel requires Flashinfer v0.6.0. @ispobock @BBuf @FlamingoPg @AniZpZ 

## Accuracy Tests

```
# launch server
python3 -m sglang.launch_server --model-path moonshotai/Kimi-K2-Thinking --tp 8 --trust-remote-code  --tool-call-parser kimi_k2 --reasoning-parser kimi_k2 --moe-runner-backend flashinfer_trtllm

# bench
python3 benchmark/gsm8k/bench_sglang.py --num-questions 2000 --parallel 2000 --num-shots 8 
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [01:02<00:00, 21.11it/s]
Accuracy: 0.937
Invalid: 0.000
Latency: 62.936 s
Output throughput: 2175.480 token/s
```

### flashinfer_trtllm mxint4 moe

```
# launch server
python3 -m sglang.launch_server --model-path moonshotai/Kimi-K2-Thinking --tp 8 --trust-remote-code  --tool-call-parser kimi_k2 --reasoning-parser kimi_k2 --moe-runner-backend flashinfer_trtllm

# perf
python3 -m sglang.bench_one_batch_server --model None --tokenizer-path xxx --base-url http://localhost:30000/ --batch-size 1 4 8 16 32 64 128 256 --input-len 8192 --output-len 1024 --show-report
```

Input lens: [8192]. Output lens: [1024].
|   batch size |   input len |   latency (s) |   input throughput (tok/s) |   output throughput (tok/s) | acc length   |   ITL (ms) |   input cost ($/1M) |   output cost ($/1M) |
|--------------|-------------|---------------|----------------------------|-----------------------------|--------------|------------|---------------------|----------------------|
|            1 |        8192 |         12.58 |                   26629.6  |                       83.45 | n/a          |      11.98 |                0.06 |                13.31 |
|            4 |        8192 |         29.2  |                   11994.3  |                      154.74 | n/a          |      25.85 |                0.13 |                 7.18 |
|            8 |        8192 |         55.95 |                    5636.69 |                      184.81 | n/a          |      43.29 |                0.28 |                 6.01 |
|           16 |        8192 |        112    |                    4820.89 |                      193.18 | n/a          |      82.82 |                0.33 |                 5.75 |
|           32 |        8192 |        211.8  |                    4571.38 |                      212.15 | n/a          |     150.84 |                0.35 |                 5.24 |
|           64 |        8192 |        414.03 |                    4215.32 |                      226.26 | n/a          |     282.86 |                0.38 |                 4.91 |
|          128 |        8192 |        720.14 |                    4124.21 |                      281.34 | n/a          |     454.97 |                0.38 |                 3.95 |

### marlin int4 moe

```
# launch server
python3 -m sglang.launch_server --model-path moonshotai/Kimi-K2-Thinking --tp 8 --trust-remote-code  --tool-call-parser kimi_k2 --reasoning-parser kimi_k2

# perf
python3 -m sglang.bench_one_batch_server --model None --tokenizer-path xxx --base-url http://localhost:30000/ --batch-size 1 4 8 16 32 64 128 256 --input-len 8192 --output-len 1024 --show-report
```

Input lens: [8192]. Output lens: [1024].
|   batch size |   input len |   latency (s) |   input throughput (tok/s) |   output throughput (tok/s) | acc length   |   ITL (ms) |   input cost ($/1M) |   output cost ($/1M) |
|--------------|-------------|---------------|----------------------------|-----------------------------|--------------|------------|---------------------|----------------------|
|            1 |        8192 |         14.45 |                   17837.9  |                       73.19 | n/a          |      13.66 |                0.09 |                15.18 |
|            4 |        8192 |         32.65 |                    4550.16 |                      160.96 | n/a          |      24.85 |                0.35 |                 6.9  |
|            8 |        8192 |         68.76 |                    4024.22 |                      156.11 | n/a          |      51.25 |                0.39 |                 7.12 |
|           16 |        8192 |        129.84 |                    3683.36 |                      173.82 | n/a          |      92.05 |                0.43 |                 6.39 |
|           32 |        8192 |        247.08 |                    3357.15 |                      193.9  | n/a          |     165.03 |                0.47 |                 5.73 |
|           64 |        8192 |        461.72 |                    3247.13 |                      218.27 | n/a          |     293.22 |                0.49 |                 5.09 |
|          128 |        8192 |        771.02 |                    3106.93 |                      302.34 | n/a          |     423.36 |                0.51 |                 3.68 |

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 1]
```diff
@@ -1200,3 +1231,310 @@ def apply(
...
+
+        layer.register_parameter("w13_weight_shape", w13_weight_shape)
+        set_weight_attrs(w13_weight_shape, extra_weight_attrs)
+
+        w2_weight_shape = torch.nn.Parameter(
+            torch.empty(num_experts, 2), requires_grad=False
+        )
+        layer.register_parameter("w2_weight_shape", w2_weight_shape)
+        set_weight_attrs(w2_weight_shape, extra_weight_attrs)
+
```
[the remark this answers is not in the document]
Comment (HandH1998) [reply python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:1900]: We can keep these for debugging and better understanding.

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 2]
```diff
@@ -42,6 +51,7 @@
 from sglang.srt.utils import (
     get_bool_env_var,
     is_cuda,
+    is_flashinfer_available,
```
Comment (ispobock) [inline python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:59]: It seems will be overrided in L80?

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 3]
```diff
@@ -67,6 +77,17 @@
     from aiter.fused_moe import fused_moe
     from aiter.ops.shuffle import shuffle_weight
 
+if is_flashinfer_available():
+    from flashinfer.fp4_quantization import block_scale_interleave
+    from flashinfer.fused_moe import (
+        convert_to_block_layout,
+        trtllm_mxint4_block_scale_moe,
```
Comment (ispobock) [inline python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:91]: What flashinfer version do we need? Is it alrady released?

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 4]
```diff
@@ -1200,3 +1231,308 @@ def apply(
             routed_scaling_factor=self.moe_runner_config.routed_scaling_factor,
         )
         return StandardCombineInput(hidden_states=output)
+
+
+class CompressedTensorsMxInt4MoEMethod(CompressedTensorsMoEMethod):
```
Comment (ispobock) [inline python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:1800]: Maybe we can update the usage in [cookbook](https://cookbook.sglang.io/docs/autoregressive/Moonshotai/Kimi-K2) for kimi-k2-thinking.

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 3 above)
Comment (HandH1998) [reply python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:91]: v0.6.0, yes it is released

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 2 above)
Comment (HandH1998) [reply python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:59]: It is a function and is called in L80.

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 3 above)
Comment (HandH1998) [reply python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:91]: We need to upgrade flashinfer to v0.6.0 then run the CI again.

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 4 above)
Comment (HandH1998) [reply python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:1800]: It seems the cookbook only provides basic usage. I'm not sure if it should be put into the cookbook.

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 3 above)
Comment (ispobock) [reply python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:91]: Yes, we may need a seperate PR to upgrade flashinfer. It's a major version update and may introduce issues.

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 4 above)
Comment (ispobock) [reply python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:1800]: Maybe we can add a note in the Configuration Tips

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 3 above)
Comment (ispobock) [reply python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:91]: https://github.com/sgl-project/sglang/pull/15551

**Comment by HandH1998** [comment]
> @HandH1998 The flashinfer is already updated to 0.6.1. Could you rebase the PR and rerun the ci for testing?

ok

**Review by ispobock — APPROVED** [review]
